File: Java/src/main/java/org/ulpgc/benchmark/BenchmarkMatrixMultiplication.java
package org.ulpgc.benchmark;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import org.openjdk.jmh.annotations.*;
import org.openjdk.jmh.results.RunResult;
import org.openjdk.jmh.runner.options.OptionsBuilder;
import org.ulpgc.matrix.basic.BasicMatrixMultiplication;
import org.ulpgc.matrix.parallel.ParallelMatrixMultiplication;
import org.ulpgc.matrix.mapreduce.MapReduceMatrixMultiplication;
import org.ulpgc.utils.MatrixGenerator;

import java.io.File;
import java.io.IOException;
import java.util.*;
import java.util.concurrent.TimeUnit;

@BenchmarkMode(Mode.AverageTime)
@OutputTimeUnit(TimeUnit.MILLISECONDS)
@Fork(1)
@Warmup(iterations = 1, time = 1, timeUnit = TimeUnit.SECONDS)
@Measurement(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
public class BenchmarkMatrixMultiplication {

    @State(Scope.Thread)
    public static class GlobalMatrixState {
        @Param({"64", "128", "256", "512", "1024"})
        public int matrixSize;

        public String inputDirPath;
        public String outputDirPath;

        public double[][] matrixA;
        public double[][] matrixB;

        @Setup(Level.Trial)
        public void setup() {
            inputDirPath = "src/main/resources/matrix_" + matrixSize + "/input";
            outputDirPath = "src/main/resources/matrix_" + matrixSize + "/output";

            matrixA = MatrixGenerator.generateDenseMatrix(matrixSize, matrixSize, 1, 10);
            matrixB = MatrixGenerator.generateDenseMatrix(matrixSize, matrixSize, 1, 10);

            MatrixGenerator.generateMatrixForMapReduce(matrixSize, matrixSize, "A", 1, 10, inputDirPath + "/matrixA.txt");
            MatrixGenerator.generateMatrixForMapReduce(matrixSize, matrixSize, "B", 1, 10, inputDirPath + "/matrixB.txt");
        }
    }

    @State(Scope.Thread)
    public static class ParallelState {
        @Param({"1", "2", "4", "8", "16"})
        public int numThreads;
    }

    @Benchmark
    public double[][] basicMultiplication(GlobalMatrixState state) {
        return new BasicMatrixMultiplication().multiply(state.matrixA, state.matrixB);
    }

    @Benchmark
    public double[][] parallelMultiplication(GlobalMatrixState matrixState, ParallelState parallelState) {
        return new ParallelMatrixMultiplication(parallelState.numThreads).multiply(matrixState.matrixA, matrixState.matrixB);
    }

    @Benchmark
    public double[][] mapReduceMultiplication(GlobalMatrixState state) {
        try {
            return MapReduceMatrixMultiplication.multiply(
                    state.inputDirPath + "/matrixA.txt",
                    state.inputDirPath + "/matrixB.txt",
                    state.outputDirPath,
                    state.matrixSize
            );
        } catch (Exception e) {
            throw new RuntimeException("Error during MapReduce multiplication", e);
        }
    }

    public static void main(String[] args) throws Exception {
        org.openjdk.jmh.runner.Runner runner = new org.openjdk.jmh.runner.Runner(new OptionsBuilder()
                .include(BenchmarkMatrixMultiplication.class.getSimpleName())
                .build());
        Collection<RunResult> results = runner.run();

        saveResultsAsJson(results);
    }

    private static void saveResultsAsJson(Collection<RunResult> results) throws IOException {
        ObjectMapper objectMapper = new ObjectMapper();
        objectMapper.enable(SerializationFeature.INDENT_OUTPUT);

        List<Map<String, Object>> allResults = new ArrayList<>();

        for (RunResult result : results) {
            Map<String, Object> resultData = new HashMap<>();
            resultData.put("Benchmark", result.getParams().getBenchmark());
            resultData.put("matrixSize", result.getParams().getParam("matrixSize"));
            resultData.put("numThreads", result.getParams().getParam("numThreads"));
            resultData.put("Mode", result.getParams().getMode().name());
            resultData.put("Cnt", result.getParams().getMeasurement().getCount());
            resultData.put("Score", result.getPrimaryResult().getScore());
            resultData.put("Error", result.getPrimaryResult().getScoreError());
            resultData.put("Units", result.getPrimaryResult().getScoreUnit());

            allResults.add(resultData);
        }

        File outputFile = new File("benchmark_results.json");
        objectMapper.writeValue(outputFile, allResults);
    }
}

File: Python/benchmark/benchmark_matrix_multiplication.py
import os
import pytest

from utils.matrix_generator import MatrixGenerator
from matrix.basic_matrix_multiplication import BasicMatrixMultiplication
from matrix.parallel_matrix_multiplication import ParallelMatrixMultiplication
from matrix.mapreduce_matrix_multiplication import DistributedMatrixMultiplication

@pytest.fixture(scope="module")
def setup_mapreduce_matrices():
    matrix_sizes = [64, 128, 256, 512, 1024]
    base_path = "benchmark_results"
    mapreduce_matrices = {}

    for size in matrix_sizes:
        input_dir = os.path.join(base_path, f"matrix_{size}", "input")
        output_dir = os.path.join(base_path, f"matrix_{size}", "output")
        os.makedirs(input_dir, exist_ok=True)
        os.makedirs(output_dir, exist_ok=True)

        matrix_a_path = os.path.join(input_dir, "MatrixA.txt")
        matrix_b_path = os.path.join(input_dir, "MatrixB.txt")
        MatrixGenerator.generate_matrix_for_mapreduce(size, size, "A", 1, 10, matrix_a_path)
        MatrixGenerator.generate_matrix_for_mapreduce(size, size, "B", 1, 10, matrix_b_path)

        mapreduce_matrices[size] = {
            "input_dir": input_dir,
            "output_dir": output_dir,
            "matrix_a_path": matrix_a_path,
            "matrix_b_path": matrix_b_path,
        }
    return mapreduce_matrices


@pytest.fixture(scope="module")
def setup_in_memory_matrices():
    matrix_sizes = [64, 128, 256, 512, 1024]
    in_memory_matrices = {}
    for size in matrix_sizes:
        matrix_a = MatrixGenerator.generate_dense_matrix(size, size)
        matrix_b = MatrixGenerator.generate_dense_matrix(size, size)
        in_memory_matrices[size] = {
            "matrix_a": matrix_a,
            "matrix_b": matrix_b,
        }
    return in_memory_matrices


@pytest.mark.parametrize("size", [64, 128, 256, 512, 1024])
def test_basic_multiplication(benchmark, setup_in_memory_matrices, size):
    matrix_a = setup_in_memory_matrices[size]["matrix_a"]
    matrix_b = setup_in_memory_matrices[size]["matrix_b"]

    def multiply():
        return BasicMatrixMultiplication().multiply(matrix_a, matrix_b)

    result = benchmark(multiply)
    assert len(result) == size and len(result[0]) == size


@pytest.mark.parametrize("size", [64, 128, 256, 512, 1024])
@pytest.mark.parametrize("num_threads", [1, 2, 4, 8, 16])
def test_parallel_multiplication(benchmark, setup_in_memory_matrices, size, num_threads):
    matrix_a = setup_in_memory_matrices[size]["matrix_a"]
    matrix_b = setup_in_memory_matrices[size]["matrix_b"]

    parallel_multiplication = ParallelMatrixMultiplication(num_threads)

    def multiply():
        return parallel_multiplication.multiply(matrix_a, matrix_b)

    result = benchmark(multiply)
    assert len(result) == size and len(result[0]) == size


@pytest.mark.parametrize("size", [64, 128, 256, 512, 1024])
def test_mapreduce_multiplication(benchmark, setup_mapreduce_matrices, size):
    config = setup_mapreduce_matrices[size]
    matrix_a_path = config["matrix_a_path"]
    matrix_b_path = config["matrix_b_path"]
    output_dir = config["output_dir"]

    def multiply():
        job = DistributedMatrixMultiplication(args=[
            f'--matrix-a={matrix_a_path}',
            f'--matrix-b={matrix_b_path}'
        ])
        job.sandbox()
        with job.make_runner() as runner:
            runner.run()

    benchmark(multiply)
